### ISSUE TITLE
Vdbe_field_ref does not use full power of built-in tuple field map

### ISSUE THREAD

**Issue by Gerold103** [tarantool/tarantool#4267]
After #3691 is pushed, there will be a problem with `vdbe_field_ref` structure. It existed even before, but afterwards it will be visible more clearly.

Vdbe field ref is a dynamic index over tuple fields storing offsets to each field and filling the offset array on demand. It is highly used in SQL, because it strongly relies on fast and repetitive access to any field, not only indexed.

There is an optimisation for the case when a requested field `fieldno` is indexed, and the tuple itself stores offset to the field in its own small field map, used by indexes. `vdbe_field_ref` then uses that map to retrieve offset value without decoding anything. But when SQL requests any field `> fieldno`, then `vdbe_field_ref` decodes the tuple from the beginning in the worst case. Even having previously accessed `fieldno`. But it could start decoding from the latter.

The optimisation is really easy to implement, but probably can give a performance impact.

**Comment by kshcherbatov** [comment]
Kostya:
```
A rename is only the first step. It's also about managing the
responsibility. As far as I can see, the new class is a proxy for
an entire tuple, not a single field. This doesn't seem convenient
going forward, but since there is a shared state (the slots array),
let it be. 

But then the name is wrong, it's vdbe_tuple_ref, not
vdbe_field_ref. Why did you blindly rename without objecting to a
another stupid name? 

The next thing is managing the responsibility for the tuple
itself. 

If we begin to think that this is a tuple ref from vdbe, then this
ref should own the tuple, and vdbe should always access the tuple
using this ref. So vdbe_field_ref_create should ref the tuple,
vdbe_field_ref_destroy() should free it. Again, please object if
it doesn't make sense - or it may require some refactoring
(likely).

Finally, the choice of the representation for the slots array and
the implementation itself. When initializing the fetcher class,
you do not consult with vdbe what tuple fields are used. Yet you
do know it from the parsing - and only select fields (one or two)
are used in most cases.

For example: 

CHECK age >=0 and age <= 170 -- sane age

So you don't need to allocate or store a lot of slots in most
cases. Moreover, perhaps it's wise to only use an optimization of
there are few slots in use, and cache Mem objects, not offsets to
fields, and if there are too many slots in use, do nothing?

Please keep in mind that the same object could be used for WHERE,
HAVING or other clauses in the future.



All in all the job turns out to be big and complex enough so that
you please feel free to extract it into a separate patch.

I do not insist, I just don't want to hold up the push.

On the other hand, I do think it's critical enough to not postpone
it any further.
```
